micro_sam/sam_annotator/_widgets.py

```diff
@@ -0,0 +1,76 @@
+from enum import Enum
+import os
+from pathlib import Path
+from typing import TYPE_CHECKING, Optional
+
+from magicgui import magic_factory, widgets
+from napari.qt.threading import thread_worker
+
+from micro_sam.sam_annotator._state import AnnotatorState
+from micro_sam.util import (
+    ImageEmbeddings,
+    get_sam_model,
+    precompute_image_embeddings,
+    _MODEL_URLS,
+    _DEFAULT_MODEL,
+    _available_devices,
+)
+
+if TYPE_CHECKING:
+    import napari
+
+Model = Enum("Model", _MODEL_URLS)
+available_devices_list = ["auto"] + _available_devices()
+
+
+@magic_factory(
+    pbar={'visible': False, 'max': 0, 'value': 0, 'label': 'working...'},
+    call_button="Compute image embeddings",
+    device = {"choices": available_devices_list},
+    save_path={"mode": "d"},  # choose a directory
+    )
+def embedding_widget(
+    pbar: widgets.ProgressBar,
+    image: "napari.layers.Image",
+    model: Model = Model.__getitem__(_DEFAULT_MODEL),
+    device = "auto",
+    save_path: Optional[Path] = None,  # where embeddings for this image are cached (optional)
+    optional_custom_weights: Optional[Path] = None,  # A filepath or URL to custom model weights.
+) -> ImageEmbeddings:
+    """Image embedding widget."""
+    # Make sure save directory exists and is an empty directory
+    if save_path is not None:
+        if not save_path.exists():
+            os.makedirs(save_path)
+        if not save_path.is_dir():
+            raise NotADirectoryError(
+                f"The user selected 'save_path' is not a direcotry: {save_path}"
+            )
+        if len(os.listdir(save_path)) > 0:
```

Comment:
I think this defeats the purpose of caching the affinities: the idea is that a user can close the annotation tool and open it again later while pointing to the same `save_path` so that the embeddings don't have to be recomputed.
This check prevents this and makes caching the embeddings not useful.

Reply:
This also relates to your comment from above:

> I've put in a bit of code to make sure the save_path directory exists and is empty. It prints an error message to the terminal if the directory is not empty.

Printing the error here is not a good idea. Maybe you intended this because of the problem with the hash comparison, but I don't want to print an error for the correct behavior of using already cached affinities.
We could use a warning instead for now and remove this once we figure out the issue with the hash.

Reply:
That's reasonable - I've changed the code to fix this problem.
Now it works in all situations: (1) creating new image embeddings, (2) returning/loading existing image embeddings if the data_signature of the save directory matches our input image, and (3) making an error popup in the napari viewer if something goes wrong (data signature does not match, save directory is not a zarr array or empty folder, anything else unexpected that raises an actual error).